File: app/src/main/java/ru/wasiliysoft/zcashnanopoolorg/Frafment/GeneralFragment.java
package ru.wasiliysoft.zcashnanopoolorg.Frafment;

import android.app.Fragment;
import android.app.LoaderManager;
import android.content.Context;
import android.content.Loader;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.io.IOException;
import java.util.List;

import retrofit2.Response;
import ru.wasiliysoft.zcashnanopoolorg.Activity.MainActivity;
import ru.wasiliysoft.zcashnanopoolorg.App;
import ru.wasiliysoft.zcashnanopoolorg.Model.NpCalc;
import ru.wasiliysoft.zcashnanopoolorg.Model.NpGeneral;
import ru.wasiliysoft.zcashnanopoolorg.NpLoader;
import ru.wasiliysoft.zcashnanopoolorg.R;

public class GeneralFragment extends Fragment implements LoaderManager.LoaderCallbacks<Response> {

    private static final int LOADER_ID_NPCALC = 1;
    private static final int LOADER_ID_NPGENERAL = 2;
    private TextView tvCurH;
    private SwipeRefreshLayout mySwipeRefreshLayout;
    private TextView tvBalance, tvUnconfirmedBalance;
    private LinearLayout llWorkers;
    private LayoutInflater ltInflater;

    private TableLayout tlCalc;
    private TableLayout tlAVGHashRate;
    private static NpGeneral.Data mNpGenData;
    private LoaderManager lm;
    private String mAddress;

    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAddress = getArguments().getString(MainActivity.BUNDLE_MINER_ADDRESS);
        // Инициализация загрузчика
        lm = getLoaderManager();
        lm.initLoader(LOADER_ID_NPGENERAL, null, this);
//        getLoaderManager().initLoader(LOADER_ID_NPCALC, null, this);
    }

    void Refresh() {

        tlAVGHashRate.removeAllViews();
        tlCalc.removeAllViews();
        llWorkers.removeAllViews();
        lm.restartLoader(LOADER_ID_NPGENERAL, null, this).forceLoad();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup parent,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_general, parent, false);

        tvCurH = v.findViewById(R.id.tvCurH);
        tvBalance = v.findViewById(R.id.tvBalance);
        tvUnconfirmedBalance = v.findViewById(R.id.tvUnconfirmedBalance);

        mySwipeRefreshLayout = v.findViewById(R.id.swiperefresh);
        mySwipeRefreshLayout.setOnRefreshListener(
                new SwipeRefreshLayout.OnRefreshListener() {
                    @Override
                    public void onRefresh() {
                        Refresh();
                    }
                }
        );

        ltInflater = getActivity().getLayoutInflater();
        tlCalc = v.findViewById(R.id.tlCalc);
        llWorkers = v.findViewById(R.id.llWorkers);
        tlAVGHashRate = v.findViewById(R.id.tlAvgHashrate);

        return v;
    }

    void settingBalances() {
        tvBalance.setText(mNpGenData.getBalance());
        tvUnconfirmedBalance.setText(mNpGenData.getUnconfirmedBalance());
    }

    void settingCurHashrate() {
        tvCurH.setText(mNpGenData.getHashrate());
    }

    void settingAVGHashRate() {
        NpGeneral.AvgHashrate h = mNpGenData.getAvgHashrate();
        tlAVGHashRate.removeAllViews();
        tlAVGHashRate.addView(getViewAVGHashRate(h.getH1(), "1 hour"));
        tlAVGHashRate.addView(getViewAVGHashRate(h.getH3(), "3 hour"));
        tlAVGHashRate.addView(getViewAVGHashRate(h.getH6(), "6 hour"));
        tlAVGHashRate.addView(getViewAVGHashRate(h.getH12(), "12 hour"));
        tlAVGHashRate.addView(getViewAVGHashRate(h.getH24(), "24 hour"));
    }

    View getViewAVGHashRate(String hashrate, String period) {
        View viewCalc = ltInflater.inflate(R.layout.avg_data, null, false);
        //                Period
        TextView c1 = viewCalc.findViewById(R.id.tvCol1);
        c1.setText(period);
        //                VGHashRate
        TextView c2 = viewCalc.findViewById(R.id.tvCol2);
        c2.setText(hashrate);
        return viewCalc;
    }

    void settingEarnings(NpCalc.Data c) {
        tlCalc.removeAllViews();
        tlCalc.addView(ltInflater.inflate(R.layout.calc_data_header, null, false));
        tlCalc.addView(getViewEarnings(c.getDay()));
        tlCalc.addView(getViewEarnings(c.getWeek()));
        tlCalc.addView(getViewEarnings(c.getMonth()));
    }

    View getViewEarnings(NpCalc.Earnings o) {

        View viewCalc = ltInflater.inflate(R.layout.calc_data, null, false);
        //                Period
        TextView c1 = viewCalc.findViewById(R.id.tvCol1);
        c1.setText(o.getPeriod());
        //                Coin
        TextView c2 = viewCalc.findViewById(R.id.tvCol2);
        c2.setText(o.getCoins());
        //                BTC
        TextView c3 = viewCalc.findViewById(R.id.tvCol3);
        c3.setText(o.getBitcoins());
        //                USD
        TextView c4 = viewCalc.findViewById(R.id.tvCol4);
        c4.setText(o.getDollars());
        //                RUR
        TextView c5 = viewCalc.findViewById(R.id.tvCol5);
        c5.setText(o.getRubles());
        return viewCalc;
    }

    void settingWorkers() {
        llWorkers.removeAllViews();
        List<NpGeneral.Worker> mListWorkers = mNpGenData.getWorkers();
        for (NpGeneral.Worker w : mListWorkers) {
            View view = ltInflater.inflate(R.layout.worker, null, false);

            ((TextView) view.findViewById(R.id.tvWorker)).setText(w.getId());

            ((TextView) view.findViewById(R.id.tvCurH))
                    .setText(new StringBuilder()
                            .append("[ Current Hashrate  ")
                            .append(w.getHashrate())
                            .append(" ]")
                    );
            ((TextView) view.findViewById(R.id.tvAVGH6))
                    .setText(new StringBuilder()
                            .append("[ AVG (6h) Hashrate ")
                            .append(w.getH6())
                            .append(" ]")
                    );

            long diffMinutes = (System.currentTimeMillis() / 1000 - w.getLastshare()) / 60;
            long hours = diffMinutes / 60;
            long minutes = diffMinutes % 60;
            ((TextView) view.findViewById(R.id.tvLastShare))
                    .setText(new StringBuilder()
                            .append("[ last share ")
                            .append(hours)
                            .append("h. ")
                            .append(minutes)
                            .append("min. ago ]")
                    );
            llWorkers.addView(view);
        }
    }

    @Override
    public void onLoadFinished(Loader<Response> loader, Response response) {
        mySwipeRefreshLayout.setRefreshing(false);
        Log.d("onLoadFinished id", loader.getId() + "");

        if (response == null || !response.isSuccessful()) {
            Toast.makeText(getActivity(), R.string.error_loading_data, Toast.LENGTH_SHORT).show();
            return;
        }
        switch (loader.getId()) {
            case LOADER_ID_NPGENERAL:
                NpGeneral.Resp mGeneralResp = (NpGeneral.Resp) response.body();
                if (!mGeneralResp.getStatus()) {
//                    TODO Обхекдинить класыы Response (выделить общее во внешний класс) для удаления дулирования кода
                    Toast.makeText(getActivity(), R.string.error_loading_data, Toast.LENGTH_SHORT).show();
                    return;
                }
                mNpGenData = mGeneralResp.getData();
                settingBalances();
                settingCurHashrate();
                settingWorkers();
                settingAVGHashRate();
//                Старт загрузки прибыльности
                if (!mNpGenData.getAvgHashrate().getH6unRound().equals("0.0")) {
                    lm.restartLoader(LOADER_ID_NPCALC, null, this).forceLoad();
                }
                break;
            case LOADER_ID_NPCALC:
                NpCalc.Resp mCalcResp = (NpCalc.Resp) response.body();
                if (!mCalcResp.getStatus()) {
                    Toast.makeText(getActivity(), R.string.error_loading_data, Toast.LENGTH_SHORT).show();
                    return;
                }
                NpCalc.Data mCalcData = mCalcResp.getData();
                settingEarnings(mCalcData);
                break;
        }

    }

    @Override
    public void onLoaderReset(Loader<Response> loader) {

    }

    @Override
    public Loader<Response> onCreateLoader(int i, Bundle bundle) {

        Log.d("onCreateLoader id", i + "");
        switch (i) {
            case LOADER_ID_NPGENERAL:
                return new GeneralLoader(getActivity(), mAddress);
            case LOADER_ID_NPCALC:
                if (mNpGenData != null) {
                    return new CalcLoader(getActivity(), mNpGenData.getAvgHashrate().getH6unRound());
                }
        }
        return null;
    }

    static class GeneralLoader extends NpLoader {
        String user;

        GeneralLoader(Context context, String user) {
            super(context);
            this.user = user;
        }

        @Override
        protected Response LoadData() throws IOException {
            Response mResp = App.getApi().getGeneral(user).execute();
            if (mResp.isSuccessful()) {
                return mResp;
            } else {
                Log.e("LoadData", " Error get General data, NP return status false");
                return null;
            }
        }
    }

    static class CalcLoader extends NpLoader {
        String hashrate;

        CalcLoader(Context context, String hashrate) {
            super(context);
            this.hashrate = hashrate;
        }

        @Override
        protected Response LoadData() throws IOException {
            Response mResp = App.getApi().getCalcCoin(hashrate).execute();
            if (mResp.isSuccessful()) {
                return mResp;
            } else {
                Log.e("LoadData", " Error get Calc data, NP return status false");
                return null;
            }
        }
    }
}

File: app/src/main/java/ru/wasiliysoft/zcashnanopoolorg/Frafment/GeneralFragment.kt
package ru.wasiliysoft.zcashnanopoolorg.Frafment


import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.work.WorkInfo
import androidx.work.WorkManager
import kotlinx.android.synthetic.main.fragment_general.*
import ru.wasiliysoft.zcashnanopoolorg.App
import ru.wasiliysoft.zcashnanopoolorg.Model.Miner
import ru.wasiliysoft.zcashnanopoolorg.Model.NpCalc
import ru.wasiliysoft.zcashnanopoolorg.Model.NpGeneral
import ru.wasiliysoft.zcashnanopoolorg.R
import ru.wasiliysoft.zcashnanopoolorg.npWorker

class GeneralFragment : Fragment() {
    private var minerId: Int = 0
    private lateinit var ltInflater: LayoutInflater

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        minerId = arguments!!.getInt(BUNDLE_MINER_ID)
    }

    fun resetViews() {
        lvBalances.visibility = View.GONE

        lvCurHashRate.visibility = View.GONE

        lvAvgHashrate.visibility = View.GONE
        lvAvgHashrate.removeAllViews()

        tlCalc.visibility = View.GONE
        tlCalc.removeAllViews()

        lvWorkers.visibility = View.GONE
        lvWorkers.removeAllViews()
    }

    fun settingsViews(ninew: Miner) {
        if (ninew.gen != null) {
            settingGeneral(ninew.gen!!)
            if (ninew.calc != null) {
                settingCalc(ninew.calc!!)
            }
        }
    }

    fun Refresh() {
        val worker1 = npWorker.newInstance(minerId)
        WorkManager.getInstance().getWorkInfoByIdLiveData(worker1.id).observe(this, Observer<WorkInfo> { worker ->
            if (worker != null) {
                Log.d(TAG_WORKER, worker.state.toString())
                when (worker.state) {
                    WorkInfo.State.RUNNING -> {
                        swiperefresh.isRefreshing = true
                        resetViews()
                    }
                    WorkInfo.State.SUCCEEDED -> {
                        tvFailedLoadData.visibility = View.GONE
                        swiperefresh.isRefreshing = false
                        settingsViews(App.getMiners().read()[minerId])
                    }
                    WorkInfo.State.FAILED -> {
                        tvFailedLoadData.visibility = View.VISIBLE
                        swiperefresh.isRefreshing = false
                        resetViews()
                    }
                }
            }
        })
        WorkManager.getInstance().enqueue(worker1)
    }

    override fun onCreateView(inflater: LayoutInflater, parent: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_general, parent, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        swiperefresh.setOnRefreshListener { Refresh() }
        ltInflater = getActivity()!!.layoutInflater
        Refresh()
    }

    fun settingGeneral(data: NpGeneral.Data) {
        tvBalance!!.text = data.balance
        tvUnconfirmedBalance!!.text = data.unconfirmedBalance
        lvBalances.visibility = View.VISIBLE

        tvCurH!!.text = data.hashrate
        lvCurHashRate.visibility = View.VISIBLE

        val h = data.avgHashrate

        lvAvgHashrate.addView(getViewAVGHashRate(h.h1, "01 hour"))
        lvAvgHashrate.addView(getViewAVGHashRate(h.h3, "03 hour"))
        lvAvgHashrate.addView(getViewAVGHashRate(h.h6, "06 hour"))
        lvAvgHashrate.addView(getViewAVGHashRate(h.h12, "12 hour"))
        lvAvgHashrate.addView(getViewAVGHashRate(h.h24, "24 hour"))
        lvAvgHashrate.visibility = View.VISIBLE

        settingWorkers(data)
    }


    fun getViewAVGHashRate(hashrate: String, period: String): View {
        val viewCalc = ltInflater.inflate(R.layout.avg_data, null, false)
        //                Period
        val c1 = viewCalc.findViewById<TextView>(R.id.tvCol1)
        c1.text = period
        //                VGHashRate
        val c2 = viewCalc.findViewById<TextView>(R.id.tvCol2)
        c2.text = hashrate
        return viewCalc
    }

    fun settingCalc(data: NpCalc.Data) {
        tlCalc.addView(ltInflater.inflate(R.layout.calc_data_header, null, false))
        tlCalc.addView(getViewEarnings(data.day))
        tlCalc.addView(getViewEarnings(data.week))
        tlCalc.addView(getViewEarnings(data.month))
        tlCalc.visibility = View.VISIBLE
    }

    fun getViewEarnings(o: NpCalc.Earnings): View {

        val viewCalc = ltInflater.inflate(R.layout.calc_data, null, false)
        //                Period
        val c1 = viewCalc.findViewById<TextView>(R.id.tvCol1)
        c1.text = o.period
        //                Coin
        val c2 = viewCalc.findViewById<TextView>(R.id.tvCol2)
        c2.text = o.coins
        //                BTC
        val c3 = viewCalc.findViewById<TextView>(R.id.tvCol3)
        c3.text = o.bitcoins
        //                USD
        val c4 = viewCalc.findViewById<TextView>(R.id.tvCol4)
        c4.text = o.dollars
        //                RUR
        val c5 = viewCalc.findViewById<TextView>(R.id.tvCol5)
        c5.text = o.rubles
        return viewCalc
    }

    fun settingWorkers(data: NpGeneral.Data) {
        val mListWorkers = data.workers
        var wCount = mListWorkers.size - 1
        val workersLimit = 100
        if (wCount > workersLimit) {
            limitWorkerNotify.text = "Sorry, max $workersLimit workers showing"
            limitWorkerNotify.visibility = View.VISIBLE
            wCount = workersLimit - 1
        } else {
            limitWorkerNotify.visibility = View.GONE
        }
        if (wCount != 0) {
            var w: NpGeneral.Worker
            var diffMinutes: Long = 0
            var hours: Long = 0
            var minutes: Long = 0
            var hashrate: String
            var h6: String
            for (i in 0..wCount) {
                w = mListWorkers[i]
                val view = ltInflater.inflate(R.layout.worker, null, false)

                diffMinutes = (System.currentTimeMillis() / 1000 - w.lastshare!!) / 60
                hours = diffMinutes / 60
                minutes = diffMinutes % 60
                hashrate = w.hashrate
                h6 = w.h6

                val tvWorker: TextView = view.findViewById<View>(R.id.tvWorker) as TextView
                tvWorker.text = "[" + (i + 1) + "] " + w.id
                if (diffMinutes >= 20) {
                    tvWorker.setTextColor(resources.getColor(R.color.offlineWorker))
                }
                (view.findViewById<View>(R.id.tvCurH) as TextView).text = "[ Current Hashrate  $hashrate ]"
                (view.findViewById<View>(R.id.tvAVGH6) as TextView).text = "[ AVG (6h) Hashrate $h6 ]"
                (view.findViewById<View>(R.id.tvLastShare) as TextView).text = "[ last share $hours h. $minutes min. ago ]"
                lvWorkers.addView(view)
            }
        }
        lvWorkers.visibility = View.VISIBLE
    }


    companion object {
        val TAG = "GeneralFragment"
        val TAG_WORKER = "GeneralFragment_Worker"
        var BUNDLE_MINER_ID = "BUNDLE_MINER_ID"

        fun newInstance(i: Int): GeneralFragment {
            val args = Bundle()
            args.putInt(GeneralFragment.BUNDLE_MINER_ID, i)
            val f = GeneralFragment()
            f.arguments = args
            return f
        }
    }
}
